include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -136,47 +236,251 @@ struct __parallel_merge_submitter<_IdType, __internal::__optional_kernel_name<_N
 
         const _IdType __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
 
-        auto __event = __exec.queue().submit([&](sycl::handler& __cgh) {
-            oneapi::dpl::__ranges::__require_access(__cgh, __rng1, __rng2, __rng3);
-            __cgh.parallel_for<_Name...>(sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id) {
-                const _IdType __i_elem = __item_id.get_linear_id() * __chunk;
-                const auto __start = __find_start_point(__rng1, __rng2, __i_elem, __n1, __n2, __comp);
-                __serial_merge(__rng1, __rng2, __rng3, __start.first, __start.second, __i_elem, __chunk, __n1, __n2,
-                               __comp);
+        auto __event = __exec.queue().submit(
+            [&__rng1, &__rng2, &__rng3, __comp, __chunk, __steps, __n1, __n2](sycl::handler& __cgh) {
+                oneapi::dpl::__ranges::__require_access(__cgh, __rng1, __rng2, __rng3);
+                __cgh.parallel_for<_MergeKernelName...>(
+                    sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id) {
+                        const _IdType __i_elem = __item_id.get_linear_id() * __chunk;
+                        const auto __start = __find_start_point(__rng1, __rng2, __i_elem, __n1, __n2, __comp);
+                        __serial_merge(__rng1, __rng2, __rng3, __start.first, __start.second, __i_elem, __chunk, __n1,
+                                       __n2, __comp);
+                    });
             });
+        // We should return the same thing in the second param of __future for compatibility
```

Comment:
It seem we should not...
`__parallel_merge_submitter::operator()` and `__parallel_merge_submitter_large::operator()`
return the same type.... 

line 252:  "future(event,  `__result_and_scratch_storage_base_ptr{}`)
line 412: future(event, __p_result_and_scratch_storage_base), 

where the second arg is also an instance of `__result_and_scratch_storage_base_ptr` type, because line 
396 tells: `__result_and_scratch_storage_base_ptr __p_result_and_scratch_storage_base;`

So, we may just writing  `return __future(__event);` in the both cases.

Reply:
Please take a look into the code, what exactly we return and which types it has:
- in the `__parallel_merge_submitter::operator()` :
```C++
return __future(__event, __result_and_scratch_storage_base_ptr{});
```
- in the `__parallel_merge_submitter_large::operator()` :
```C++
return __future(std::move(__event), std::move(__p_result_and_scratch_storage_base));
```

So what is your question?

Reply:
About your suggestion "So, we may just writing return __future(__event); in the both cases." - it's absolutely incorrect approach. Please read the description of this PR.

Reply:
Discussed offline.

Reply:
yes, sorry for my mistake..
__p_result_and_scratch_storage_base is defined on line 402 (with other meaning), not 396...

Reply:
No problems.